crates/ark/src/lsp/main_loop.rs

```diff
@@ -692,3 +676,191 @@ impl std::fmt::Debug for TraceKernelNotification<'_> {
+///   retaining only the most recent one.
+///
+/// Ideally we'd process indexer tasks continually without making them dependent
+/// on diagnostics tasks. The current setup blocks the queue loop while
+/// diagnostics are running, but it has the benefit that rounds of diagnostic
+/// refreshes don't race against each other. The frontend will receive all
+/// results in order, ensuring that diagnostics for an outdated version are
+/// eventually replaced by the most up-to-date diagnostics.
+async fn process_indexer_queue(mut rx: mpsc::UnboundedReceiver<IndexerQueueTask>) {
+    while let Some(task) = rx.recv().await {
+        // Drain all available tasks
+        let mut tasks = vec![task];
+        while let Ok(next_task) = rx.try_recv() {
+            tasks.push(next_task);
+        }
+
+        // Separate by type
+        let mut diagnostics_batch = Vec::new();
+        let mut indexer_batch = Vec::new();
+
+        for task in tasks {
+            match task {
+                IndexerQueueTask::Indexer(indexer_task) => indexer_batch.push(indexer_task),
+                IndexerQueueTask::Diagnostics(diagnostic_task) => {
+                    diagnostics_batch.push(diagnostic_task)
+                },
+            }
+        }
+
+        // Process indexer tasks first so diagnostics tasks work with an up-to-date index
+        process_indexer_batch(indexer_batch).await;
+        process_diagnostics_batch(diagnostics_batch).await;
+    }
+}
+
+async fn process_indexer_batch(batch: Vec<IndexerTask>) {
+    // Deduplicate tasks by key. We use a `HashMap` so only the last insertion
+    // is retained. `Update` tasks use URI as key, `Start` tasks use None (we
+    // only expect one though). This is effectively a way of cancelling `Update`
+    // tasks for outdated documents.
+    let batch: std::collections::HashMap<_, _> = batch
+        .into_iter()
+        .map(|task| match &task {
+            IndexerTask::Update { uri, .. } => (Some(uri.clone()), task),
+            IndexerTask::Start { .. } => (None, task),
+        })
+        .collect();
+
+    let mut handles = Vec::new();
+
+    for (_, task) in batch {
+        handles.push(tokio::task::spawn_blocking(move || match task {
+            IndexerTask::Start { folders } => {
+                indexer::start(folders);
+            },
+            IndexerTask::Update { document, uri } => {
+                let result = if let Ok(path) = uri.to_file_path() {
+                    indexer::update(&document, &path)
+                } else {
+                    Err(anyhow!("Failed to convert URI to file path: {uri}"))
+                };
+                if let Err(err) = result {
+                    log::error!("Indexer update failed: {err}");
+                }
+            },
+        }));
+    }
+
+    for handle in handles {
+        let _ = handle.await;
+    }
+}
+
+async fn process_diagnostics_batch(batch: Vec<RefreshDiagnosticsTask>) {
```

Comment:
Reconsider batching? Based on zoom call

Reply:
I take it back, the interleaved indexer/diagnostics processing will arise only if the queue tasks are processed faster than they arrive so the current setup is fine. If we get a bunch of tasks very rapidly, we will collect them, split them by type, and process the indexer tasks first. So the batching is still useful.

This setup will be entirely replaced by Salsa dependencies. Diagnostics tasks will be cancelled automatically as document updates arrive. So we shouldn't worry about this temporary setup too much.

Reply:
I adapted the loop so that we check back for more indexer tasks once we have finished a round of indexing. We do that at most 10 times so diagnostics get refreshed once in a while if the user is writing too fast to keep up. I think I'm happy with the queue setup now!